### ISSUE TITLE
Eagerly Awaiting the Megatron Code Release

### ISSUE THREAD

**Issue by mactavish91** [MoonshotAI/Moonlight#1]
I really appreciate the amazing work you’ve done on the Moonshot AI open source MoE model. I’m excited about the upcoming release of the Megatron code and can't wait to explore it.
Thank you for your efforts!

**Comment by toothacher17** [comment]
Thanks for the interest! We will publish it ASAP!

**Comment by toothacher17** [comment]
hi, @mactavish91 Please see the example proof of concept Megatron-LM PR in here:

https://github.com/NVIDIA/Megatron-LM/pull/1428

**Comment by huyiwen** [comment]
Are there any plab to open source the training code for MoE mentioned in the paper?

**Comment by toothacher17** [comment]
@huyiwen MoE training code is not related with the Muon optimizer itself, so we do not plan to release it.

**Comment by huyiwen** [comment]
are the released intermediate checkpoints for the MoE model or the dense model? If they are for the MoE model, how should we load and train them?

**Comment by toothacher17** [comment]
> are the released intermediate checkpoints for the MoE model or the dense model? If they are for the MoE model, how should we load and train them?

They are MoE model:
- for loading, using huggingface transfomrers library: https://huggingface.co/moonshotai
- for training, it is discussed in this issue: https://github.com/MoonshotAI/Moonlight/issues/11

**Comment by mactavish91** [comment]
@toothacher17 
Hello,

I've recently been experimenting with the Muon optimizer on larger models. Specifically, dense models are performing noticeably better than MoE models.

For the MoE models, I implemented a loss-free method to balance the expert workload. However, I've noticed that when using the Muon optimizer, the load imbalance across experts appears to be more pronounced compared to when using Adam.

I suspect that this increased load imbalance with Muon is a primary contributing factor to the poorer performance of the MoE models.

Do you have any insights on this observation or any suggestions for mitigating this potential load imbalance issue when using Muon with MoE architectures?

![Image](https://github.com/user-attachments/assets/8514965e-8068-4d4c-a9cd-c8c87c91efde)

**Comment by toothacher17** [comment]
hi, @mactavish91 , thanks for sharing.

First of all, moonlight model and some other internal models are trained on Muon with auxfree bias, so we do not observe that Moe + Muon + auxfree bias has some specific disadvantages in terms of balance or performances.

For your case, it seems that for both AdamW and Muon, the balance seems very bad. We usually observe a max_violates that is smaller than 0.1~0.2, while you have ~3+ max violates. For our implementation of auxfree, we made some slight modifications and you can see the appendix C of the paper:

<img width="758" alt="Image" src="https://github.com/user-attachments/assets/a7e52005-09ab-421e-96d6-231a8c1c5189" />

For the hyper parameters of balance, **we use 1e-3 score update ratio, and 1e-4 micro aux loss**, following deepseek v3. Besides, **please make sure that your auxfree bias is manually updated**, rather than updated by Muon or AdamW

**Comment by mactavish91** [comment]
> hi, [@mactavish91](https://github.com/mactavish91) , thanks for sharing.
> 
> First of all, moonlight model and some other internal models are trained on Muon with auxfree bias, so we do not observe that Moe + Muon + auxfree bias has some specific disadvantages in terms of balance or performances.
> 
> For your case, it seems that for both AdamW and Muon, the balance seems very bad. We usually observe a max_violates that is smaller than 0.1~0.2, while you have ~3+ max violates. For our implementation of auxfree, we made some slight modifications and you can see the appendix C of the paper:
> 
> <img alt="Image" width="758" src="https://private-user-images.githubusercontent.com/7029845/419818519-a7e52005-09ab-421e-96d6-231a8c1c5189.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyNTM0MDcsIm5iZiI6MTc0MTI1MzEwNywicGF0aCI6Ii83MDI5ODQ1LzQxOTgxODUxOS1hN2U1MjAwNS0wOWFiLTQyMWUtOTZkNi0yMzFhOGMxYzUxODkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDZUMDkyNTA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzZiNzdiYTIxZmIxOGQ4NGQ1NjhmZTI0YTE4ZTE4ZTRmMjc5MTkyOTAwN2IxMzQxMzhkMjM3YzQ1MTgwZDQ4NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.L6V80dK2cU8P7r6RXjR5UcodG8RGe_4cM_M89vUVZo8">
> For the hyper parameters of balance, **we use 1e-3 score update ratio, and 1e-4 micro aux loss**, following deepseek v3. Besides, **please make sure that your auxfree bias is manually updated**, rather than updated by Muon or AdamW

This metric value is around 0.2 for a 5B parameter model with a 64 expert (top-8) MoE.

However, when scaling to a 246B parameter model with a 160 expert (top-6, like DeepSeek V2) MoE, the value jumps to 3-4.

Is this normal for such a scale increase?

**Comment by toothacher17** [comment]
We trained larger scale MoE as well (not exactly as DSV3), and the balance is fine. You might want to check your dataloader or batch size? If your dataloader does not generate diverse enough data or your bsz is too small, you might have very bad balance.

Besides, please check the auxfree update rate, it is a bit tricy as well. For a sanity check, you can try not to use auxfree, but use a traditional aux loss and see if your model can balance

**Comment by mactavish91** [comment]
I tried the methods you mentioned, but the max vio is still quite high. I want to confirm if our calculation method is the same.

Here's how we calculate it:
```
num_tokens, num_experts = probs.shape
num_total_tokens = num_tokens * topk * num_sub_seq

probs_mean_per_expert = probs.mean(dim=0)
token_portions_per_expert = tokens_per_expert / num_total_tokens

torch.distributed.all_reduce(
    token_portions_per_expert, op=torch.distributed.ReduceOp.AVG, group=parallel_state.get_data_modulo_expert_parallel_group()
)

balanced_token_num = torch.tensor((num_total_tokens / num_experts,), device=tokens_per_expert.device, dtype=torch.float32)
load_gap = (token_portions_per_expert * num_total_tokens - balanced_token_num) / balanced_token_num
max_vio = load_gap.abs().max()

expert_balance_loss = expert_balance_factor * num_experts * torch.sum(probs_mean_per_expert * token_portions_per_expert)
```

Could you please confirm if our calculation method is consistent with yours?

**Comment by toothacher17** [comment]
@mactavish91 

Auxfree is usually highly related with your customized implementation. It might be hard to directly look at this piece and judge if it has some bugs or not...

One simpler way to debug is to not use auxfree for now, and use a 1e-2 or 5e-3 micro batch aux loss, and see if your model can balance. Auxfree is usually well implemented in all kinds of open-source training frameworks

**Comment by toothacher17** [comment]
hey, @mactavish91 

Our calculation might be slightly different, but it seem to both follow the definition from deepseek. An easier to understand metric is the min_ratio, which is the minimal expert tokens ratio. For example, for a moe with 64 selective experts, the expected ratio is 1/64 ~= 0.015625, and we usually observe that empirical min_ratio is about 0.012~0.013. You can see if your min expert ratio is as expected.

**Comment by mactavish91** [comment]
> hey, [@mactavish91](https://github.com/mactavish91)
> 
> Our calculation might be slightly different, but it seem to both follow the definition from deepseek. An easier to understand metric is the min_ratio, which is the minimal expert tokens ratio across all layers. For example, for a moe with 64 selective experts, the expected ratio is 1/64 = 0.015625, and we usually observe that empirical min_ratio is about 0.012 to 0.013. You can see if your min expert ratio is as expected.

I found that max_vio was being measured at the micro-batch level. After performing an all-reduce operation across all ranks, this value turned out to be around 0.1, which seems reasonable. However, I still observe a significant performance gap where the Muon optimizer outperforms the MoE model on dense models by a large margin. In my experiments with varying model scales, Muon consistently outpaced Adam by approximately 50% in terms of step ratio on dense models, whereas for MoE models, the improvement was only between 5% to 20%.

**Comment by SeunghyunSEO** [comment]
hi guys, good to see you again! 
im currently reproducing deepseek-v3 lite with muon and adam but in my setup, it's throughput is horrible (even in adam setup).
i think load balancing is problem (I'm monitoring maxvio too), but I'm not sure my distributed training setup is optimal, 
so if you don't mind, i'd happy if you guys share distributed training setup (EP/PP/VPP/mbsz/gbsz...and so on) or any advice!
@toothacher17 @mactavish91

**Comment by toothacher17** [comment]
> > hey, [@mactavish91](https://github.com/mactavish91)
> > Our calculation might be slightly different, but it seem to both follow the definition from deepseek. An easier to understand metric is the min_ratio, which is the minimal expert tokens ratio across all layers. For example, for a moe with 64 selective experts, the expected ratio is 1/64 = 0.015625, and we usually observe that empirical min_ratio is about 0.012 to 0.013. You can see if your min expert ratio is as expected.
> 
> I found that max_vio was being measured at the micro-batch level. After performing an all-reduce operation across all batches, this value turned out to be around 0.1, which seems reasonable. However, I still observe a significant performance gap where the Muon optimizer outperforms the MoE model on dense models by a large margin. In my experiments with varying model scales, Muon consistently outpaced Adam by approximately 50% in terms of step ratio on dense models, whereas for MoE models, the improvement was only between 5% to 20%.

hi, @mactavish91 We actually also trained on MoE, and as you can see from the intermediate checkpoint report: https://github.com/MoonshotAI/Moonlight/blob/master/Moonlight_intermediate_checkpoints.pdf

![Image](https://github.com/user-attachments/assets/b78f2256-bd9f-4063-b7f7-3e004dbf751d)

Besides, in the paper of comparing moonlight v.s. moonlight-A, I think muon also has big advantages:

![Image](https://github.com/user-attachments/assets/30b0b0ad-d513-48ff-80eb-35a0f2da4c91)

But it would be interesting to know your training scale, e.g. like what model size are you referring to, or how many tokens have you trained?

**Comment by toothacher17** [comment]
> deepseek-v3 lite

Hi @SeunghyunSEO what is size of your deepseek-v3 lite model? Is it with a similar size of Moonlight? Here are some details about how we train our Moonlight model, we actually saw a pretty good tflops/mfu:

1. Overall, we trained Moonlight with 5.7T tokens on a N256 H800 cluster in less than 3 days (more than 2T tokens per day);

2. [No EP, PP4VP1, No recomputes] For parallelism, we do not use EP so we will not worry about balancing issues. Instead, we used pp4vp1 to save memory and used selective recomputes on rmsnorm and mlp-swiglu. As a result, we do not need recomputes. PP is suitable for this case as it has limited bubble and since it is 27 layers, the output lm head can take 1 layer's workload to further reduce bubble;

3. [Group MM kernel] Since it is very fine-grained experts so each GPU matmal is very fast. Besides, the model has so many experts so your cpu will need to launch so many GPU matmal kernels, leading to a case that your CPU launching kernel speed becoming the bottleneck. Therefore, you'll need a group mm kernel, so you will reorganize your input for all your experts, then only launch matmal kernel three times (for the three matmal in MLP);

4. [MTP workload] Be careful with MTP. We do not use MTP for moonlight. If you want to use MTP, each MTP actually takes two layers workload so you'll need to manage it for a low bubble ratio. For example, for DSV3, it has 61 layers, so they reserve 1 layer for lm head, 1 layer for addition MTP layer, and another 1 layer for the MTP lm head. So in total, it is 64 layers workload, and can be perfectly partitioned into PPs;

5. [FAV3] We observed ~10% speedup via bumping from FAV2 to FAV3 (I think they recently supported head_dim 192) 

Let me know if you want to know more details, we are happy to share!

**Comment by SeunghyunSEO** [comment]
thank you for the fast and kind reply as always! 
you're right, I'm testing actually moonlight scale model where activated param is 2.3B and total 15.7B excluding embedding.
MLA + auxfree + shared moe + finegrained, and so on..
i used EP=4, PP=2(VPP2) but it makes sense this scale MoE maybe more scalable in PP.
i will test it and comeback soon, thank you again!

**Comment by SeunghyunSEO** [comment]
@toothacher17 
i quickly test PP prefered setup, but throughput is still low... hmm...

- A100x80GB dgx 4nodes 
- mbsz: 1, gbsz: 256, seqlen: 4k (so 1M per iter)
- so num micro batch is 32
- adamw
- EP:1, PP:4 (vpp is 1, that is one layer per stage)
- dropless
- TE grouped gemm 
- no recompute, only flash attn
- no MTP

<img width="1235" alt="Image" src="https://github.com/user-attachments/assets/cbd4146c-71d7-4938-9807-91dd121377f0" />

**Comment by toothacher17** [comment]
hi, @SeunghyunSEO Thanks for sharing. Let me think about it:

1. Thanks for following our parallelism design, I think it is a reasonable design for now. Your pp bubble = (pp-1)/acc/vp_stages = 3/32/4= ~2%, and you do not need recompute because of PP. Besides, the memory seems healthy. So in theory, you mfu should not be bounded by the parallelism design;

2. Since the overall parallelism is not bad, you'll need to check where is your bottleneck. For example, make sure that lm head stands alone as a single layer to balance the workload. Besides, you might need to check if your group MM kernel's performances are actually working as expected. An easy way to check is increase 4k to 8k, or use a larger micro batch size see if it performs better

Debugging and improving the performances of such a small model can be challenging, to further analysis, you might need to tools like ndtimleine (https://github.com/volcengine/veScale/tree/main/vescale/ndtimeline) to analyze your PP bubble, and nsys profile (https://developer.nvidia.com/nsight-systems) to further analyze.

**Comment by SeunghyunSEO** [comment]
> hi, @SeunghyunSEO Thanks for sharing. Let me think about it:
> 
> 1. Thanks for following our parallelism design, I think it is a reasonable design for now. Your pp bubble = (pp-1)/acc/vp_stages = 3/32/4= ~2%, and you do not need recompute because of PP. Besides, the memory seems healthy. So in theory, you mfu should not be bounded by the parallelism design;
> 
> 2. Since the overall parallelism is not bad, you'll need to check where is your bottleneck. For example, make sure that lm head stands alone as a single layer to balance the workload. Besides, you might need to check if your group MM kernel's performances are actually working as expected. An easy way to check is increase 4k to 8k, or use a larger micro batch size see if it performs better
> 
> Debugging and improving the performances of such a small model can be challenging, to further analysis, you might need to tools like ndtimleine (https://github.com/volcengine/veScale/tree/main/vescale/ndtimeline) to analyze your PP bubble, and nsys profile (https://developer.nvidia.com/nsight-systems) to understand CPU and GPU's performances.
> 
> 
> 
> 

well, i tried so hard within few days, but couldnt raise throughput that much. still 30++ mfu. i profiled bubble frac and groupedgemm kerenl performance and found groupedgemm could be problemtic in small mbsz setup so i increase ep but mfu was still not satisfying...
btw do you have any specific kernel for MLA?? i found deepseek use specialized kernel that fuse rope+rms+upproj+flash and current megatron's mla cause huge activation memory compared to mha or gqa and large kernel launch overhead.

**Comment by toothacher17** [comment]
> performance and found groupedgemm could be problemtic in small mbsz setup so i increase ep but mfu was still not satisfying...

Yeah this is a known issue that such small moe model will have performance issues. Using grouped gemm could help you avoid cpu launching kernel bottleneck, but the gemm itself might still be slow. Please try to use longer seqlens like 8k or use larger micro batch size, so the kernel performance might be slightly better. You can try to calculate the tflops performances only for the gemm kernel and see how it performs.


> btw do you have any specific kernel for MLA?? i found deepseek use specialized kernel that fuse rope+rms+upproj+flash and current megatron's mla cause huge activation memory compared to mha or gqa and large kernel launch overhead.

We also observed that attention caused some delay, because when we upgraded from FAV2 to FAV3, we observed end to end tflops increased by 7% to 10% for different cases. If you are using A-series card, you might not be able to use FAV3 though. I am not sure about how current MGLM implements MLA, but this is a good way to use FAV2 to speedup MLA

https://huggingface.co/deepseek-ai/DeepSeek-V2/blob/main/modeling_deepseek.py#L907

We did not implement specific MLA kernels, but used a implementation similar to the above link. It is basically similar activation memory overhead with a FAV2 (smaller if using qlora + large hidden size)

**Comment by toothacher17** [comment]
@SeunghyunSEO 

Does mglm MLA use its own dot-product-attention or flash-attn? Maybe the easiest way to speed up is to put the lora down, up projection, and rope stuff into a torch compile func, and the prepared q,k,v can use the flash-attn, like in this file:

https://huggingface.co/deepseek-ai/DeepSeek-V2/blob/main/modeling_deepseek.py#L907

**Comment by SeunghyunSEO** [comment]
@toothacher17 i think MLA without FAV3 performance is horrible (no recomputation, no fused, no qk=192,v=128 is allowed).
so i need to write a kernel or something i guess.
thank you again for everything you've done for me :)

**Comment by toothacher17** [comment]
@SeunghyunSEO  Yeah, your analysis seems reasonable. We did find that FAV3 improves performances a lot, and in the long run, we'll need such a fused kernel as well for extreme optimization. But in general, I think bumping up to FAV3 solves most efficiency concerns for us.

It would be great if you can share your profile, some of my colleagues can take a look and see if they have some ideas!

**Comment by toothacher17** [comment]
Hey @SeunghyunSEO , I talked with some other related Moonshot folks, and they are happy to help. [tangcisi](https://github.com/tangcisi) might be able to help when he gets time later.

Hey, [tangcisi](https://github.com/tangcisi), you can contact @SeunghyunSEO at [X](https://x.com/SeunghyunSEO7) for more detailed performance bmk results and help analyze how to improve it.

Cheers!
